### ISSUE TITLE
Orphan should be false if element is present

### ISSUE THREAD

**Issue by diesl** [IGreatlyDislikeJavascript/bootstrap-tourist#28]
In contrast to Bootstrap Tour, when I set global option `orphan: true`, then all steps are shown as orphans wether the element is present or not.

In my opinion, steps with a valid and present element should be shown with respective placement whereas steps without a valid element should be shown as orphans.

Also, it would be desireable to differentiate between intended orphans with `element: null` (e.g. a welcome to this tour message) and unintended orphans, where the given element is not existing.

**Comment by IGreatlyDislikeJavascript** [comment]
Interesting point, thank you. The original Tour purpose of the orphan option was to still show a step as an orphan, if that step's element was not present, exactly as you describe.

To me, that seemed illogical - if you've created a tour step against an element, it makes sense to show it against the element. If the element doesn't exist, there's an issue with the tour flow that must be handled.

That's why I added the onElementUnavailable and delayOnElement options, because I couldn't see a case where you would want to still show a tour step against a missing element.

Currently, the orphan option also always overrides the step.element option, as you've discovered.

I could envisage logic that says "for step N, show this content if element is visible, or this other content if element isn't visible", or even "here's some generic content that applies whether the element is visible or not". However that's the precise purpose of onElementUnavailable and the .goTo functions - to ensure every step is specific to the requirement.

Am I missing a use case here?

I think what you're asking is:

1. reverse the priority order: if(step.element is valid && step.orphan == true), ignore step.orphan
2. if step.element == null, always assume step.orphan == true


What I'm unclear on are these 2:

> whereas steps without a valid element should be shown as orphans
> unintended orphans, where the given element is not existing

Isn't a step with an "unintended orphan" the same as "an error in the tour steps"?

**Comment by diesl** [comment]
Reading your comment, I now see your intention. Just to make clear what I meant, I will summarize my current setup:

> Isn't a step with an "unintended orphan" the same as "an error in the tour steps"?

For steps like the mentioned welcome screen, I do not have and do not intend to bind it to a specific element. I call this _intended orphan_, in contrast to an _unintended orphan_ like a mistyped selector, for example. I used `element: ''` or something similar, and `global.orphan: true` and things were fine. Maybe it also worked when I omitted the `element` option, not sure right now.

But I can of course fix it with a new `step.orphan: true` for each of these _intended orphans_. Maybe this is indeed the most clean solution.

What I mixed in here is the part that changed: The global orphan option. I do not know if it is useful to set all steps to orphans even if the elements are valid. Is this a common need? Maybe I just find the original solution more intuitive, maybe there are usecases to show a step even if the element could not be resolved?

> reverse the priority order: if(step.element is valid && step.orphan == true), ignore step.orphan

Not sure what you mean here, how to ignore `step.orphan` if you test against it in the if-clause?

> if step.element == null, always assume step.orphan == true

Basically that's my current setup for these _intended orphans_. But to say it again, maybe it is better to use `orphan: true` on these steps, because I am not sure if a mistyped selector also returns `null` sometimes and as a result we cannot differentiate between intended or unintended.

Phew. I suppose now things are more unclear than before ;)

**Comment by IGreatlyDislikeJavascript** [comment]
I've written 3 or 4 replies and deleted them all, they didn't make sense and I finally realized something - the naming of global orphan option is wrong. In original Tour, global orphan and step orphan have similar but subtly different results.


> I call this intended orphan, in contrast to an unintended orphan

Good terminology, let's use this from now on.


> But I can of course fix it with a new step.orphan: true for each of these intended orphans

> a mistyped selector also returns null sometimes and as a result we cannot differentiate between intended or unintended.

That was exactly my intention of the step.orphan implementation. I felt specifically setting `step.orphan == true`, rather than relying on a null or invalid element (as per your comment), was good because it:

(a) made it clear in code that the step was an intended orphan
(b) ensured that `onElementUnavailable` only triggered on an unintended orphan

> I do not know if it is useful to set all steps to orphans even if the elements are valid. 

My thinking which led us here was the exact opposite! I couldn't see it being useful to show an unintended orphan (bug in tour steps) as per original Tour approach, therefore I reworked the global orphan option to force all steps to orphan after implementing `onElementUnavailable()` etc. But I accept your point that "all are orphans" may well be a fairly useless option :)

> Not sure what you mean here, how to ignore step.orphan if you test against it in the if-clause?

Currently the orphan option is prioritized. If global or per step `orphan == true`, Tourist ignores `step.element`. This is what's got us into confusion...

I think the tour global option `orphan` is wrongly named and is confusing. In **original** Tour, at global options level, `orphan` means "if a step is an _unintended_ orphan, show it as an _intended_ orphan". However at the step level, `step.orphan` means "this step is an intended orphan no matter what".

In Tourist, the difference is that the global orphan option forces all steps to intended orphans.

To fix this I believe we need the following instead:

- Remove/ignore the global bool option `orphan`. Issue: backwards drop-in compatibility with original Tour.

- Retain the bool option `step.orphan`. When `true`, this step is an intended orphan. No element is required in the step options. This makes it clear when a step is an intended orphan. (FYI, this is the current approach).

- Add global bool option `allStepsAreIntendedOrphansSoIgnoreAllStepElements`. When true, every step in the tour is shown as an intended orphan. Elements are ignored, `delayOnElement `is never called, `onElementUnavailable` is never called.

- Add global and per step bool option `showStepAsAnIntendedOrphanIfItIsAnUnintendedOrphan`. When true, if a step has an element specified which is an unintended orphan (broken element), the step will be shown as an orphan. If `delayOnElement` is specified, the delay must timeout first. The `onElementUnavailable `callback will not be called.

Does this look like a possible approach/solution?

**Comment by diesl** [comment]
Sorry for letting you wait, I forgot to answer.

Thanks for your detailed summary of the problem and ideas so far.

You could do it like that. However, apart from these really long option names 😉 I would suggest the following:

- Keep the global bool option orphan and use the same behaviour like the step orphans. This way, it uses the same logic (which is why you changed it), and can be overriden at step level, if needed. Backwards compatibility is not given, but it isn't given for the current approach too.

- `allStepsAreIntendedOrphansSoIgnoreAllStepElements` would not be needed then

- Use showStepAsAnIntendedOrphanIfItIsAnUnintendedOrphan like you described and rename to `showOnlyIfElementExists`. Well, the name is still a bit to long for my taste ...

Did I forget something important?

**Comment by IGreatlyDislikeJavascript** [comment]
No probs, I don't think anyone here does this full time, I certainly have a day job ;-)

Your suggestions make sense. 2 questions. First:

> Use `showStepAsAnIntendedOrphanIfItIsAnUnintendedOrphan `like you described and rename to `showOnlyIfElementExists`

I think this should be called `showIfUnintendedOrphan` (`bool`), which I think is a better description of what we want?

Second: what's the logical behavior when a step uses `delayOnElement`? 

The `delayOnElement` can wait for either the step.element or an arbitrary DOM element to appear before either showing the step, or timing out and calling `onElementUnavailable`.

If `delayOnElement` is set to wait for an arbitrary element, and step.element exists, do we show the step attached to step.element or as an orphan? 

Does this scenario even qualify as an unintended orphan and therefore do we consider `showIfUnintendedOrphan`? If step.element exists, therefore it's not an unintended orphan, but delayOnElement has the side effect of making it an almost-orphan because something else (nominally required for the step) doesn't exist.
 
Regardless of the above, `showIfUnintendedOrphan == true` will have to override and never call `onElementUnavailable`

Thanks for your feedback on this - I want to get the logic right because it will be a pain to unpick it later.

**Comment by IGreatlyDislikeJavascript** [comment]
I've added showIfUnintendedOrphan to the unstable/new-backdrop branch, as that will take over the master branch shortly.

**Comment by IGreatlyDislikeJavascript** [comment]
v0.3.0 uploaded into master branch, feature available!

**Comment by diesl** [comment]
I'm sorry, I forgot to answer in a timely manner again. But it is never too late to say thank you for fixing this issue.
